trace-telemetry-exporter/src/main.rs

```diff
@@ -0,0 +1,236 @@
+        &args.broker,
+        &args.username,
+        &args.password,
+    )
+    .set("group.id", &args.consumer_group)
+    .set("enable.partition.eof", "false")
+    .set("session.timeout.ms", "6000")
+    .set("enable.auto.commit", "false")
+    .create()?;
+
+    consumer.subscribe(&[&args.trace_topic])?;
+
+    let builder = PrometheusBuilder::new();
+    builder
+        .with_http_listener(args.metrics_address)
+        .install()
+        .expect("prometheus metrics exporter should be setup");
+
+    metrics::describe_counter!(
+        METRIC_MSG_COUNT,
+        metrics::Unit::Count,
+        "Number of messages received"
+    );
+
+    metrics::describe_counter!(
+        METRIC_MSG_RATE,
+        metrics::Unit::CountPerSecond,
+        "Rate of messages received"
+    );
+
+    metrics::describe_gauge!(
+        METRIC_LAST_MSG_TIMESTAMP,
+        metrics::Unit::Nanoseconds,
+        "Timestamp of last message received (ns)"
+    );
+
+    metrics::describe_gauge!(
+        METRIC_LAST_MSG_FRAME_NUMBER,
+        metrics::Unit::Count,
+        "Frame number of last message received"
+    );
+
+    metrics::describe_gauge!(
+        METRIC_CHANNEL_COUNT,
+        metrics::Unit::Count,
+        "Number of channels in last message received"
+    );
+
+    metrics::describe_gauge!(
+        METRIC_SAMPLE_COUNT,
+        metrics::Unit::Count,
+        "Number of samples in last message received"
+    );
+
+    let recent_msg_counts: MessageCounts = Arc::new(Mutex::new(HashMap::new()));
+
+    task::spawn(update_message_rate(
+        Arc::clone(&recent_msg_counts),
+        args.message_rate_interval,
+    ));
+
+    poll_kafka_msg(consumer, Arc::clone(&recent_msg_counts)).await;
+
+    Ok(())
+}
+
+fn get_digitiser_label(digitiser_id: u8) -> (String, String) {
+    ("digitiser_id".to_string(), format!("{}", digitiser_id))
+}
+
+async fn update_message_rate(recent_msg_counts: MessageCounts, message_rate_interval: u64) {
+    loop {
+        // Wait a set period of time before calculating average.
+        sleep(Duration::from_secs(message_rate_interval)).await;
+        let mut recent_msg_counts = recent_msg_counts.lock().unwrap();
+        // Calculate and record message rate for each digitiser.
+        recent_msg_counts
+            .iter()
+            .for_each(|(digitiser_id, recent_msg_count)| {
+                let msg_rate = *recent_msg_count as f64 / message_rate_interval as f64;
+                let labels = [get_digitiser_label(*digitiser_id)];
+                gauge!("digitiser_message_received_rate", &labels).set(msg_rate);
+            });
+        // Reset recent message count for each digitiser.
+        recent_msg_counts.clear();
+    }
+}
+
+fn process_message(data: &DigitizerAnalogTraceMessage<'_>, recent_msg_counts: MessageCounts) {
+    let id = data.digitizer_id();
+    let labels = [get_digitiser_label(id)];
+
+    // Increment recent message count for the digitiser
+    recent_msg_counts
```

Comment:
Has this been tested? Pretty sure this will result in no entries being added to the recent counts as you are not handling the case where they key does not exist. (I admit I have not actually tested this either yet, only picking this up from code review)

(see https://doc.rust-lang.org/std/collections/hash_map/enum.Entry.html#method.or_insert)

Reply:
Could I ask you to test this before making it ready for review again.

Reply:
Just from using a `println!` statement to print out the hashmap, it seems to be working correctly. Are there any other parts of the program that I should test?

Reply:
I mean testing the entire program, i.e. run it, send some dummy trace data, `curl` the metrics endpoint and you should see metrics and see them change as new trace data arrives.

Reply:
Thank you, it didn't occur to me that I could use `curl` to test this.